File: Canny.py
import cv2
import matplotlib.pyplot as plt
import numpy as np


def flip180(arr):
    new_arr = arr.reshape(arr.size)
    new_arr = new_arr[::-1]
    new_arr = new_arr.reshape(arr.shape)
    return new_arr

def twodConv(f, W,method='zero'):
    print(f.shape)
    h,w = f.shape
    W = flip180(W)
    k1,k2 = W.shape
    p1,p2 = int((k1-1)/2),int((k2 - 1) / 2)
    padded_img = np.zeros((h+2*p1,w+2*p2),dtype = f.dtype)
    result_img = np.zeros_like(f)
    padded_img[p1:h+p1,p2:w+p2] = f
    if method == 'zero':
        padded_img = padded_img
    elif method == 'replicate':
        padded_img[0:p1,p2:p2+w] = f[0,:]
        padded_img[p1+h:p1*2+h,p2:p2+w] = f[h-1,:]
        for i in range(p2):
            padded_img[p1:p1+h,i] = f[:,0]
            padded_img[p1:p1+h,p2+w+i] = f[:,w-1]
        padded_img[0:p1,0:p2] =  f[0,0]
        padded_img[0:p1,p2+w:p2*2 +w] = f[0,w-1]
        padded_img[p1+h:p1*2 +h,0:p2] =  f[h-1,0]
        padded_img[p1+h:p1*2 +h,p2+w:p2*2 +w] = f[h-1,w-1]
    for i in range(0,h):
        for j in range(0,w):
            result_img[i,j] = np.sum(padded_img[i:k1+i,j:j+k2] * W)
    return result_img

def gaussKernel(sig,m=None):#m=(3×σ)*2 + 1
    if m is None:
        m = round(3*sig)*2 + 1
    if m < round(3*sig)*2 + 1:
        raise("warning:m is too small")
    if sig <=0:
        sig = 0.3*((m-1)*0.5 - 1) + 0.8
    result = np.zeros((m,m),np.float32)
    for i in range(m):
        for j in range(m):
            d = np.sqrt((i-int(m/2))**2 + (j-int(m/2))**2)
            result[i,j] = np.exp(-d**2/(2*sig**2))/(2*np.pi*sig**2)
    sum_m = np.sum(result)
    result = result/sum_m
    return result




def cal_grad(img):
    sobelx = np.array([[-1,0,1],[-2,0,2],[-1,0,1]])
    sobely = np.array([[-1,-2,1],[0,0,0],[1,2,1]])
    h,w = img.shape
    gradx = np.zeros_like(img)
    grady = np.zeros_like(img)
    grads = np.zeros_like(img)
    thetas = np.zeros_like(img)
    where_grad = np.zeros_like(img)
    p1,p2 = 1,1
    padded_img = np.zeros((h+2*p1,w+2*p2),dtype = img.dtype)
    padded_img[p1:h+p1,p2:w+p2] = img
    
    for i in range(1,h+1):
        for j in range(1,w+1):
            temp = padded_img[i-1:i+2,j-1:j+2]
            temp_gradx = np.sum(temp*sobelx)
            temp_grady = np.sum(temp*sobely)
            gradx[i-1,j-1] = temp_gradx
            grady[i-1,j-1] = temp_grady
            grads[i-1,j-1] = np.sqrt(temp_gradx**2 + temp_grady**2)#callculate the grads
            theta = np.arctan(temp_gradx/temp_grady)#calculate theta
            thetas[i-1,j-1] = theta
            if theta >0 and theta <= np.pi / 4 :
                where_grad[i-1,j-1] = 0
            elif theta > np.pi/4 and theta <= np.pi/2:
                where_grad[i-1,j-1] = 1
            elif theta >= -np.pi/2 and theta <= -np.pi/4:
                where_grad[i-1,j-1] = 2
            elif theta > -np.pi/4 and theta < 0:
                where_grad[i-1,j-1] = 3

    return grads,thetas,where_grad

def rgb1gray(f,method='NTSC'):
    rows,cols,c = f.shape
    dtype_of_f = f.dtype
    grey_img = np.zeros((rows,cols),dtype=dtype_of_f)
    if method == 'NTSC':
        grey_img = f[:,:,0]*0.1140 + f[:,:,1]*0.5870 + f[:,:,2] * 0.2989#BGR
    elif method == 'average':
        grey_img = (f[:,:,0] + f[:,:,1] + f[:,:,2])/3
    else:
        raise("you should write the correct method")
    return grey_img

def NMS(img,where_grad,thetas,grads,lowthres):
    h,w = img.shape
    highvalue = np.max(img)
    lowvalue = lowthres * highvalue
    result = np.zeros_like(img)
    for i in range(1,h-1):
        for j in range(1,w-1):
            N= img[i-1,j]
            S = img[i+1,j]
            W = img[i,j-1]
            E = img[i,j+1]
            NW = img[i-1,j-1]
            NE = img[i-1,j+1]
            SW = img[i+1,j-1]
            SE = img[i+1,j+1]
            case ,theta = where_grad[i,j] ,thetas[i,j]
            tantheta = np.tan(theta)
            if case == 0:
                gp1 , gp2 = (1-tantheta) *E + tantheta * NE ,(1-tantheta) *W + tantheta * SW
            elif case == 1:
                gp1 , gp2 = (1-tantheta) *N + tantheta * NE ,(1-tantheta) *S + tantheta * SW
            elif case == 2:
                gp1 , gp2 = (1-tantheta) *N + tantheta * NW ,(1-tantheta) *S + tantheta * SE
            elif case == 3:
                gp1 , gp2 = (1-tantheta) *W + tantheta * NW ,(1-tantheta) *E + tantheta * SE

            if grads[i,j] >=gp1 and grads[i,j] >=gp2:
                if grads[i,j] >= highvalue:
                    grads[i,j] = highvalue
                    result[i,j] = 255

                elif grads[i,j] < lowvalue:
                    grad[i,j] = 0
                else:
                    grads[i,j] = lowvalue
            else:
                grads[i,j] = 0
        
    for i in range(1,h-1):
        for j in range(1,w-1):
            if grads[i,j] == lowvalue:
                temps = grads[i-1:i+2,j-1:j+2]
                if np.any(temps==highvalue):
                    result[i,j] = 255

    
    return result

if __name__ == "__main__":
    print("start")
    img_path = "lena512color.tiff"
    img = cv2.imread(img_path)
    img = rgb1gray(img)
    plt.imshow(img,cmap='gray')
    plt.show()
    print("read succesfull")
    highvalue = np.max(img)
    lowthres = 0.05
    GW = gaussKernel(sig=1)
    flitered_img = twodConv(img,GW)
    grads,thetas,where_grad = cal_grad(flitered_img)
    results = NMS(flitered_img,where_grad,thetas,grads,lowthres)
    plt.imshow(results,cmap='gray')
    plt.show()

File: Canny.cpp
#include<opencv2/opencv.hpp>
#include<stdarg.h>
using namespace cv;
const double PI = 3.1415926;
#define CV_SORT_EVERY_ROW    0
#define CV_SORT_EVERY_COLUMN 1
#define CV_SORT_ASCENDING    0
#define CV_SORT_DESCENDING   16



Mat conv2D(cv:: Mat img,int kSize,cv:: Mat kernel) {
    int row = img.rows;
    int col = img.cols;
    cv::Rect rect;
    Mat src;
    src = img.clone();
    printf("kernel类型：%d，img类型:%d\n",kernel.type(),src.type());
   
    Mat filterImg = Mat::zeros(row, col, CV_64FC1);
    int gaussCenter = kSize / 2;
    for (int i = gaussCenter; i < row - gaussCenter; i++) {
        for (int j = gaussCenter; j < col - gaussCenter; j++) {
            rect.x = j - gaussCenter;
            rect.y = i - gaussCenter;
            rect.width = kSize;
            rect.height = kSize;
            //printf("******************\n");
            printf("%d %d\n", i, j);
           
            Mat temp = src(rect);
           
            filterImg.at<double>(i, j) = cv::sum(kernel.mul(src(rect))).val[0];
            
        }
    }
    return filterImg;
}

void calGrad(cv:: Mat img,Mat* whereGrad1,Mat* grad1,Mat* thetas1) {
    int row = img.rows;
    int col = img.cols;
    cv::Rect temp;
    Mat gradX = Mat::zeros(row, col, CV_64FC1); //水平梯度
    Mat gradY = Mat::zeros(row, col, CV_64FC1); //垂直梯度
     
    Mat grad = *grad1;//梯度幅值
    
    Mat thetas = *thetas1;//梯度角度
    
    Mat whereGrad = *whereGrad1;//标记不同区域
     //x方向的Sobel算子
    Mat Sx = (cv::Mat_<double>(3, 3) << -1, 0, 1, -2, 0, 2, -1, 0, 1);
    //y方向的Sobel算子
    Mat Sy = (cv::Mat_<double >(3, 3) << 1, 2, 1, 0, 0, 0, -1, -2, -1);
    for (int i = 1; i < row - 1; i++) {
        for (int j = 1; j < col - 1; j++) {
            temp.x = j - 1;
            temp.y = i - 1;
            temp.width = 3;
            temp.height = 3;
            Mat rectImg = Mat::zeros(3, 3, CV_64FC1);
            img(temp).copyTo(rectImg);
            //梯度和角度
            gradX.at<double>(i, j) += cv::sum(rectImg.mul(Sx)).val[0];
            gradY.at<double>(i, j) += cv::sum(rectImg.mul(Sy)).val[0];
            grad.at<double>(i, j) = sqrt(pow(gradX.at<double>(i, j), 2) + pow(gradY.at<double>(i, j), 2));
            thetas.at<double>(i, j) = atan(gradY.at<double>(i, j) / gradX.at<double>(i, j));
            if (0 <= thetas.at<double>(i, j) <= (PI / 4.0)) {
                whereGrad.at<double>(i, j) = 0;
            }
            else if (PI / 4.0 < thetas.at<double>(i, j) <= (PI / 2.0)) {
                whereGrad.at<double>(i, j) = 1;
            }
            else if (-PI / 2.0 <= thetas.at<double>(i, j) <= (-PI / 4.0)) {
                whereGrad.at<double>(i, j) = 2;
            }
            else if (-PI / 4.0 < thetas.at<double>(i, j) < 0) {
                whereGrad.at<double>(i, j) = 3;
            }
        }
    }
    double gradMax;
    cv::minMaxLoc(grad, &gradMax);
    if (gradMax != 0) {
        grad = grad / gradMax;
    }
    *grad1 = grad;
    *whereGrad1 = whereGrad;
    *thetas1 = thetas;

}

Mat NMS(cv:: Mat img, Mat grad, Mat whereGrad,Mat thetas,double highThres,double lowThres) {
    int row = img.rows;
    int col = img.cols;
    Mat result = Mat::zeros(row, col, CV_64FC1);
    cv::Mat caculateValue = cv::Mat::zeros(row, col, CV_64FC1); //grad变成一维
    resize(grad, caculateValue, Size(1, grad.rows * grad.cols));
    cv::sort(caculateValue, caculateValue, CV_SORT_EVERY_COLUMN + CV_SORT_ASCENDING);//升序
    long long highIndex = row * col * highThres;
    double highValue = caculateValue.at<double>(highIndex, 0); //最大阈值
    double lowValue = highValue * lowThres;
    //NMS
    for (int i = 1; i < row - 1; i++) {
        for (int j = 1; j < col - 1; j++) {
            // 八个方位
            double N = grad.at<double>(i - 1, j);
            double NE = grad.at<double>(i - 1, j + 1);
            double E = grad.at<double>(i, j + 1);
            double SE = grad.at<double>(i + 1, j + 1);
            double S = grad.at<double>(i + 1, j);
            double SW = grad.at<double>(i - 1, j - 1);
            double W = grad.at<double>(i, j - 1);
            double NW = grad.at<double>(i - 1, j - 1); // 区域判断，线性插值处理
            double tanThead; // tan角度
            double Gp1; // 两个方向的梯度强度
            double Gp2; // 求角度，绝对值
            tanThead = abs(tan(thetas.at<double>(i, j)));
            //线性插值
            switch ((int)whereGrad.at<double>(i, j)) {
            case 0: Gp1 = (1 - tanThead) * E + tanThead * NE; Gp2 = (1 - tanThead) * W + tanThead * SW; break;
            case 1: Gp1 = (1 - tanThead) * N + tanThead * NE; Gp2 = (1 - tanThead) * S + tanThead * SW; break;
            case 2: Gp1 = (1 - tanThead) * N + tanThead * NW; Gp2 = (1 - tanThead) * S + tanThead * SE; break;
            case 3: Gp1 = (1 - tanThead) * W + tanThead * NW; Gp2 = (1 - tanThead) * E + tanThead * SE; break;
            default: break;
            }
            // NMS -非极大值抑制和双阈值检测
            if (grad.at<double>(i, j) >= Gp1 && grad.at<double>(i, j) >= Gp2) {
                //双阈值检测
                if (grad.at<double>(i, j) >= highValue) {
                    grad.at<double>(i, j) = highValue;
                    result.at<double>(i, j) = 255;
                }
                else if (grad.at<double>(i, j) < lowValue) {
                    grad.at<double>(i, j) = 0;
                }
                else {
                    grad.at<double>(i, j) = lowValue;
                }
            }
            else {
                grad.at<double>(i, j) = 0;
            }
        }
    }
    //抑制孤立低阈值点 3*3. 找到高阈值就255
    cv::Rect temp;
    for (int i = 1; i < row - 1; i++) {
        for (int j = 1; j < col - 1; j++) {
            if (grad.at<double>(i, j) == lowValue) {
                //3*3 区域强度
                temp.x = i - 1;
                temp.y = j - 1;
                temp.width = 3;
                temp.height = 3;
                for (int x = 0; x < 3; x++) {
                    for (int y = 0; y < 3; y++) {
                        if (grad(temp).at<double>(x, y) == highValue) {
                            result.at<double>(i, j) = 255;
                            break;
                        }
                    }
                }
            }
        }
    }
    return result;
}


Mat getKernel(int kSize) {
    int gaussCenter = kSize / 2;
    double  sigma = 1;
    Mat guassKernel = Mat::zeros(kSize, kSize, CV_64FC1);
    for (int i = 0; i < kSize; i++) {
        for (int j = 0; j < kSize; j++) {
            guassKernel.at<double>(i, j) = (1.0 / (2.0 * PI * sigma * sigma)) * (double)exp(-(((double)pow((i - (gaussCenter + 1)), 2) + (double)pow((j - (gaussCenter + 1)), 2)) / (2.0 * sigma * sigma)));
        }
    }
    Scalar sumValueScalar = cv::sum(guassKernel);
    double sum = sumValueScalar.val[0];
    guassKernel = guassKernel / sum;
    return guassKernel;
}



int main()
{
    Mat girl = imread("lena512color.tiff"); //载入图像到Mat
    
    cvtColor(girl, girl, COLOR_BGR2GRAY);
    int co = girl.cols;
    int ro = girl.rows;
    int ch = girl.channels();
    printf("%d %d %d", co, ro, ch);
    imshow("女孩头像", girl);//显示名为 "【1】动漫图"的窗口  
    waitKey(0);
    Mat gussKernel = getKernel(3);
    printf("getkernel***********\n");
    girl.convertTo(girl, 6);
    Mat filterImg = conv2D(girl, 3, gussKernel);
    printf("conv2d*********\n");
    Mat grad = Mat::zeros(ro, co, CV_64FC1); //梯度幅值
    Mat thetas = Mat::zeros(ro, co, CV_64FC1); //梯度角度
    Mat whereGrad = Mat::zeros(ro, co, CV_64FC1);//区域
    calGrad(filterImg, &whereGrad, &grad, &thetas);
    printf("calGrad*********\n");
    Mat results = NMS(girl, grad, whereGrad, thetas, 0.9, 0.05);
    imshow("边缘检测结果", results);
    waitKey(0);
    imwrite("results3.bmp", results);


    return 0;
}